### ISSUE TITLE
Improved encoding and decoding speed of Vec<u8>

### ISSUE THREAD

**Issue by dtolnay** [bincode-org/bincode#665]
I noticed this comment:

https://github.com/bincode-org/bincode/blob/a22afa384d02ae0653d27eb0908da6771baecb26/src/de/impls.rs#L449-L452

You can implement the condition soundly as `non_static_type_id::<T>() == TypeId::of::<u8>()` where:

```rust
use std::any::TypeId;
use std::marker::PhantomData;
use std::mem;

fn non_static_type_id<T: ?Sized>() -> TypeId {
    trait NonStaticAny {
        fn get_type_id(&self) -> TypeId where Self: 'static;
    }

    impl<T: ?Sized> NonStaticAny for PhantomData<T> {
        fn get_type_id(&self) -> TypeId where Self: 'static {
            TypeId::of::<T>()
        }
    }

    let phantom_data = PhantomData::<T>;
    NonStaticAny::get_type_id(unsafe {
        mem::transmute::<&dyn NonStaticAny, &(dyn NonStaticAny + 'static)>(&phantom_data)
    })
}
```

This is sound because all types with the same TypeId as `u8` are `u8`. In contrast, not all types with the same TypeId as `&'static str` are `&'static str`.

**Comment by VictorKoenders** [comment]
cc #618 

Will implement this, thanks

Just to confirm, this is still sound to call with types with a lifetime, e.g. if I make a function like:
```rs
fn type_equal<T1, T2>() -> bool {
    non_static_type_id::<T1>() == non_static_type_id::<T2>()
}

// this is fine:
type_equal::<&'static str, u8>();
type_equal::<u8, &'static str>();

// this could result in unsoundness:
// type_equal::<&'a str, &'b str>();
```

**Comment by dtolnay** [comment]
It is sound to _get_ the TypeId for any T using this method. You only need to be concerned about the soundness of unsafe code that downcasts based on the value of the obtained TypeId, as different types can have the same TypeId. So `non_static_type_id::<T>() == TypeId::of::<&'static str>()` does not mean that you can downcast to `&'static str`.

**Comment by VictorKoenders** [comment]
Implemented this in #667. I ended up making a small helper crate for this: 
- https://github.com/bincode-org/unty
- https://docs.rs/unty
- https://crates.io/crates/unty
